File: c++/1_two_sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

vector<int> twoSum(vector<int> &nums, int target) {
    unordered_map<int, int> map;
    vector<int> ans;
    for (int i = 0; i < nums.size(); ++i) {
        if (map.count(nums[i])) {
            ans.push_back(map[nums[i]]);
            ans.push_back(i);
            break;
        }
        map[target - nums[i]] = i;
    }
    return ans;
}

int main() {
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    vector<int> ans = twoSum(nums, target);
    cout << ans[0] << " " << ans[1] << endl;
}

File: py/1_Two_Sum.py
from typing import List


def twoSum(nums: List[int], target: int) -> List[int]:
    hashmap = {}
    for i in range(len(nums)):
        temp = target - nums[i]
        if temp not in hashmap:
            hashmap[nums[i]] = i
        else:
            return [hashmap[temp], i]


def test_twoSum():
    # unit test twoSum
    nums = [2, 7, 11, 15]
    target = 9
    assert twoSum(nums, target) == [0, 1]
    print("Pass")
    print("---------------------")


def test_twoSum2():
    # unit test twoSum
    nums = [3, 2, 4]
    target = 6
    assert twoSum(nums, target) == [1, 2]
    print("Pass")
    print("---------------------")
